### ISSUE TITLE
PERL Net::OSCAR

### ISSUE THREAD

**Issue by rhandorf** [mk6i/retro-aim-server#28]
Hello again!

I'm trying to re-write a good ol fashioned PERL bot using Net::OSCAR (https://metacpan.org/pod/Net::OSCAR). 

I've edited the Common.pl file to reference the correct host, and tcpdump shows traffic. When I run the debug tool:
oscartest --loglevel=9

the output is below:

Screenname: abc
Password: 
login: Connecting to [snip]:5190.
login: State changed (FD 3) to write
login: Got r=0, w=1, e=0
login: Connected.
login: State changed (FD 3) to read
login: Got r=1, w=0, e=0
login: Getting connack.
login: Got connack.
login: Got connack.  Sending connack.
login: Connected to login server.
login: Put SNAC 0x0017/0x0006: initial_signon_request
login: SNAC time.
login: Got r=1, w=0, e=0
login: Got SNAC 0x0017/0x0007: authentication_key
login: Sending password.
login: Put SNAC 0x0017/0x0002: signon
login: Got r=1, w=0, e=0
login: Got SNAC 0x0017/0x0003: authorization_response
login: Login OK - connecting to BOS
basic OSCAR service: Connecting to [snip]:5191.
basic OSCAR service: State changed (FD 4) to write
login: State changed (FD 3) to deleted
login: Closing.
basic OSCAR service: Got r=0, w=1, e=0
basic OSCAR service: Connected.
basic OSCAR service: State changed (FD 4) to read
basic OSCAR service: Got r=1, w=0, e=0
basic OSCAR service: Getting connack.
basic OSCAR service: Got connack.
basic OSCAR service: Sending BOS-Signon.
basic OSCAR service: Put SNAC 0x0000/0x0001: BOS_signon
basic OSCAR service: SNAC time.

The server has the following event code:

time=2024-05-19T19:47:42.163Z level=DEBUG msg="accepted connection" svc=BOS ip=[snip]
time=2024-05-19T19:47:42.183Z level=INFO msg="user session failed" svc=BOS err="unexpected EOF"

Kind of curious as to what thoughts you may have in order to get Net::OSCAR working.

Thanks!

**Comment by rhandorf** [comment]
An alternative question is, are ya'll aware of any other code libraries, modules, etc that would allow me to make an AIM bot that works with this?

Thanks!

**Comment by mk6i** [comment]
> An alternative question is, are ya'll aware of any other code libraries, modules, etc that would allow me to make an AIM bot that works with this?

RAS exports all the types necessary to get a chatbot working in golang. I was actually working on a sample application this past weekend. I expect to have it ready this weekend or next. Be on the lookout for that.

In terms of the PERL app, I'll have to dig into this a bit.

**Comment by mk6i** [comment]
For those following at home, an example chat bot that uses Retro AIM Server packages is available here: https://github.com/mk6i/smarter-smarter-child

**Comment by rhandorf** [comment]
Also as an update, I pulled the latest code and the Perl module now seems to move further:

[truncated the previous data]

basic OSCAR service: Put SNAC 0x0000/0x0001: BOS_signon
basic OSCAR service: SNAC time.
basic OSCAR service: Got r=1, w=0, e=0
basic OSCAR service: Got EOF on socket
basic OSCAR service: State changed (FD 4) to deleted
basic OSCAR service: Closing.
Fatal error 0 in basic OSCAR service: Lost connection to BOS

I strangely dont see any server trace logs when connecting and it timing out.

**Comment by mk6i** [comment]
Can you please share the perl file? I’ll connect to my local server and run the debugger to see what’s up.

**Comment by rhandorf** [comment]
To install it, type:

perl -MCPAN -e shell

If you havent set up CPAN, there's some defaults to accept. Then type (case sensitive)

install Net::OSCAR

It'll pull and install the modules. Type 'exit' to get out of CPAN.

You need to manually edit one of the Perl modules to point to the OSCAR host.

Edit /usr/local/share/perl/5.30.0/Net/OSCAR/Common.pm, on or about line number 193.

Then you should be able to just run, natively, the command:
oscartest --loglevel=9

**Comment by mk6i** [comment]
**Notes**

Ran perl in a docker container:

```sh
docker run -it --network host perl /bin/bash
```

Command to run `oscartest` with hostname:

```sh
oscartest --loglevel=9 --host 192.168.68.61
```

**Comment by mk6i** [comment]
Net::OSCAR failed to work with RAS because it assumed a hard-coded auth cookie length of 256 bytes in the login FLAP frame, whereas the RAS auth cookie length was variable. When connecting to BOS, Net::OSCAR put the auth cookie in a TLV with length 256 and the cookie, which was less than 256 bytes. When the server allocated 256 bytes to read the cookie, it came up short and errored out with an unexpected end-of-file error.

I've implemented a simple solution that pads the auth cookie so that it's always 256 bytes. I feel that this hack is worthwhile because other 3rd party clients might also work this way, and it appears to be more faithful to how AIM used to work.

**Comment by mk6i** [comment]
```sh
root@docker-desktop:/usr/src/app# oscartest --loglevel=9 --host 192.168.68.61
Screenname: mk6i
Password:
login: Connecting to 192.168.68.61:5190.
login: State changed (FD 3) to write
login: Got r=0, w=1, e=0
login: Connected.
login: State changed (FD 3) to read
login: Got r=1, w=0, e=0
login: Getting connack.
login: Got connack.
login: Got connack.  Sending connack.
login: Connected to login server.
login: Put SNAC 0x0017/0x0006: initial_signon_request
login: SNAC time.
login: Got r=1, w=0, e=0
login: Got SNAC 0x0017/0x0007: authentication_key
login: Sending password.
login: Put SNAC 0x0017/0x0002: signon
login: Got r=1, w=0, e=0
login: Got SNAC 0x0017/0x0003: authorization_response
login: Login OK - connecting to BOS
basic OSCAR service: Connecting to 192.168.68.61:5191.
basic OSCAR service: State changed (FD 4) to write
login: State changed (FD 3) to deleted
login: Closing.
basic OSCAR service: Got r=0, w=1, e=0
basic OSCAR service: Connected.
basic OSCAR service: State changed (FD 4) to read
basic OSCAR service: Got r=1, w=0, e=0
basic OSCAR service: Getting connack.
basic OSCAR service: Got connack.
basic OSCAR service: Sending BOS-Signon.
basic OSCAR service: Put SNAC 0x0000/0x0001: BOS_signon
basic OSCAR service: SNAC time.
basic OSCAR service: Got r=1, w=0, e=0
basic OSCAR service: Got SNAC 0x0001/0x0003: server_ready
basic OSCAR service: Put SNAC 0x0001/0x0017: set_service_versions
basic OSCAR service: Put SNAC 0x0001/0x0006: rate_info_request
basic OSCAR service: Got r=1, w=0, e=0
basic OSCAR service: Got SNAC 0x0001/0x0018: host_versions
basic OSCAR service: Got r=1, w=0, e=0
basic OSCAR service: Got SNAC 0x0001/0x0007: rate_info_response
basic OSCAR service: Put SNAC 0x0001/0x0008: rate_acknowledgement
basic OSCAR service: BOS handshake complete!
basic OSCAR service: Signon BOS handshake complete!
basic OSCAR service: Put SNAC 0x0001/0x000E: personal_info_request
basic OSCAR service: Put SNAC 0x0013/0x0002: buddylist_rights_request
basic OSCAR service: Put SNAC 0x0013/0x0004: buddylist_request
basic OSCAR service: Put SNAC 0x0002/0x0002: locate_rights_request
basic OSCAR service: Put SNAC 0x0003/0x0002: buddy_rights_request
basic OSCAR service: Put SNAC 0x0004/0x0004: IM_parameter_request
basic OSCAR service: Put SNAC 0x0009/0x0002: BOS_rights_request
basic OSCAR service: Got r=1, w=0, e=0
basic OSCAR service: Got SNAC 0x0001/0x000F: self_information
basic OSCAR service: Got r=1, w=0, e=0
basic OSCAR service: Got SNAC 0x0013/0x0003: buddylist_3_response
basic OSCAR service: Got r=1, w=0, e=0
basic OSCAR service: Got SNAC 0x0013/0x0006: buddylist
OSCAR session: Got BLI entry  0x0001/0x0000/0x0000 with 20 bytes of data:
	  00 C8 00 02  3A 30                                      .... :0
OSCAR session: Got BLI entry Friends 0x0001/0x3A30/0x0000 with 20 bytes of data:
	  00 C8 00 04  13 AD 03 03                                .... ....
OSCAR session: Got BLI entry asd123 0x0000/0x3A30/0x0303 with 20 bytes of data:
OSCAR session: Got BLI entry mike 0x0000/0x3A30/0x13AD with 20 bytes of data:
basic OSCAR service: Put SNAC 0x0004/0x0002: add_IM_parameters
basic OSCAR service: Put SNAC 0x0001/0x0002: set_tool_versions
OSCAR session: Setting extended status.
basic OSCAR service: Put SNAC 0x0001/0x001E: set_extended_status
basic OSCAR service: Put SNAC 0x0001/0x0011: set_idle
basic OSCAR service: Put SNAC 0x0013/0x0007: buddylist_done
You are now signed on to AOL Instant Messenger.
basic OSCAR service: Got r=1, w=0, e=0
basic OSCAR service: Got SNAC 0x0002/0x0003: locate_rights_response
basic OSCAR service: Got r=1, w=0, e=0
basic OSCAR service: Got SNAC 0x0003/0x0003: buddy_rights_response
basic OSCAR service: Got r=1, w=0, e=0
basic OSCAR service: Got SNAC 0x0004/0x0005: IM_parameter_response
basic OSCAR service: Got r=1, w=0, e=0
basic OSCAR service: Got SNAC 0x0009/0x0003: BOS_rights_response
OSCAR session: Setting user information.
basic OSCAR service: Put SNAC 0x0002/0x0004: set_info
basic OSCAR service: Got r=1, w=0, e=0
basic OSCAR service: Got SNAC 0x0001/0x000F: self_information
basic OSCAR service: Got r=1, w=0, e=0
basic OSCAR service: Got SNAC 0x0003/0x000B: buddy_status_update
basic OSCAR service: Incoming bogey - er, I mean buddy - mike
OSCAR session: Got a capability.
OSCAR session: Got capability fileshare.
OSCAR session: Got a capability.
OSCAR session: Got capability sendlist.
OSCAR session: Got a capability.
OSCAR session: Got capability chat.
OSCAR session: Got a capability.
OSCAR session: Got capability filexfer.
OSCAR session: Got a capability.
OSCAR session: Got capability directim.
OSCAR session: Got a capability.
OSCAR session: Got capability buddyicon.
OSCAR session: Got a capability.
OSCAR session: Got capability addins.
Got buddy mike from Friends
send mike hello
basic OSCAR service: Put SNAC 0x0004/0x0006: outgoing_IM
basic OSCAR service: Got r=1, w=0, e=0
basic OSCAR service: Got SNAC 0x0004/0x000C: IM_acknowledgement
Your message, 547701, was sent to mike.
```

**Comment by mk6i** [comment]
Released in v0.6.0, feel free to reopen if need be.
